src/librustc_mir/transform/const_prop.rs

```diff
@@ -560,6 +545,32 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                     return None;
                 }
             }
+
+            Rvalue::Aggregate(_, operands) if operands.len() == 0 => {
+                // FIXME(wesleywiser): const eval will turn this into a `const Scalar(<ZST>)` that
+                // `SimplifyLocals` doesn't know it can remove.
+                return None;
+            }
+
+            Rvalue::Aggregate(box AggregateKind::Adt(_, variant_idx, _, _, _), _) => {
+                match &place_layout.variants {
+                    Variants::Single { index } => {
+                        if index != variant_idx {
+                            trace!("skipping uninhabited variant {:?}", rvalue);
+                            return None;
+                        }
+                    }
+                    Variants::Multiple { variants, .. } => {
+                        let variant_layout = &variants[*variant_idx];
+                        if variant_layout.abi == Abi::Uninhabited {
```

Comment:
I think this is wrong. We should never special-case uninhabited enums/variants. Rather, the behavior for uninhabited enums/variants should fall out of the general proper treatment of enums.

Empty enums are not magic or special in any way, and at least in Miri, we historically had some checks for empty enums that were very wrong because they meant we would not detect the same kinds of bugs for e.g. 0-variant enums and 2-variant enums.

@wesleywiser why did you add this, and what happens if you remove it? IMO there's a bug elsewhere if this is needed.

Reply:
You're right, this branch isn't strictly needed.

Reply:
I suppose this is the branch you added for better surface-language level consistence (the enum discussion we had). But when viewed at the MIR level, as explained yesterday, I think this introduces *inconsistency*.

So, I'd be curious what the MIR and TyLayout details look like in the two cases that you compared (where you observed inconsistent behavior that this branch fixes).

Reply:
So for both `enum Test { A(u8), B(Empty) }` and `enum Test { A(u8), B(Empty), C }`, the MIR in question is the same:

```
    bb2: {
        StorageLive(_3);
        _3 = move ((_1 as Some).0: Empty);
        StorageLive(_4);
        StorageLive(_5);
        _5 = move _3;
        ((_4 as B).0: Empty) = move _5;
        discriminant(_4) = 1;
        StorageDead(_5);
        StorageDead(_4);
        StorageDead(_3);
        goto -> bb3;
    }
```

[playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0b9d6673577a02a59b0db160bf9b6fcc)

Reply:
<details>
<summary>
Layout details for <code>enum Test1 { A(u8), B(Empty) }</code> 
</summary>

```
TyLayout {
        ty: Test1,
        details: LayoutDetails {
            variants: Single {
                index: 0,
            },
            fields: Arbitrary {
                offsets: [
                    Size {
                        raw: 0,
                    },
                ],
                memory_index: [
                    0,
                ],
            },
            abi: Scalar(
                Scalar {
                    value: Int(
                        I8,
                        false,
                    ),
                    valid_range: 0..=255,
                },
            ),
            largest_niche: None,
            align: AbiAndPrefAlign {
                abi: Align {
                    pow2: 0,
                },
                pref: Align {
                    pow2: 3,
                },
            },
            size: Size {
                raw: 1,
            },
        },
    }
```

</details>

<details>

<summary>
Layout details for <code>enum Test2 { A(u8), B(Empty), C }</code>
</summary>

```
TyLayout {
        ty: Test2,
        details: LayoutDetails {
            variants: Multiple {
                discr: Scalar {
                    value: Int(
                        I8,
                        false,
                    ),
                    valid_range: 0..=2,
                },
                discr_kind: Tag,
                discr_index: 0,
                variants: [
                    LayoutDetails {
                        variants: Single {
                            index: 0,
                        },
                        fields: Arbitrary {
                            offsets: [
                                Size {
                                    raw: 1,
                                },
                            ],
                            memory_index: [
                                0,
                            ],
                        },
                        abi: Aggregate {
                            sized: true,
                        },
                        largest_niche: None,
                        align: AbiAndPrefAlign {
                            abi: Align {
                                pow2: 0,
                            },
                            pref: Align {
                                pow2: 3,
                            },
                        },
                        size: Size {
                            raw: 2,
                        },
                    },
                    LayoutDetails {
                        variants: Single {
                            index: 1,
                        },
                        fields: Arbitrary {
                            offsets: [
                                Size {
                                    raw: 1,
                                },
                            ],
                            memory_index: [
                                0,
                            ],
                        },
                        abi: Uninhabited,
                        largest_niche: None,
                        align: AbiAndPrefAlign {
                            abi: Align {
                                pow2: 0,
                            },
                            pref: Align {
                                pow2: 3,
                            },
                        },
                        size: Size {
                            raw: 1,
                        },
                    },
                    LayoutDetails {
                        variants: Single {
                            index: 2,
                        },
                        fields: Arbitrary {
                            offsets: [],
                            memory_index: [],
                        },
                        abi: Aggregate {
                            sized: true,
                        },
                        largest_niche: None,
                        align: AbiAndPrefAlign {
                            abi: Align {
                                pow2: 0,
                            },
                            pref: Align {
                                pow2: 3,
                            },
                        },
                        size: Size {
                            raw: 1,
                        },
                    },
                ],
            },
            fields: Arbitrary {
                offsets: [
                    Size {
                        raw: 0,
                    },
                ],
                memory_index: [
                    0,
                ],
            },
            abi: ScalarPair(
                Scalar {
                    value: Int(
                        I8,
                        false,
                    ),
                    valid_range: 0..=2,
                },
                Scalar {
                    value: Int(
                        I8,
                        false,
                    ),
                    valid_range: 0..=255,
                },
            ),
            largest_niche: Some(
                Niche {
                    offset: Size {
                        raw: 0,
                    },
                    scalar: Scalar {
                        value: Int(
                            I8,
                            false,
                        ),
                        valid_range: 0..=2,
                    },
                },
            ),
            align: AbiAndPrefAlign {
                abi: Align {
                    pow2: 0,
                },
                pref: Align {
                    pow2: 3,
                },
            },
            size: Size {
                raw: 2,
            },
        },
    }
```

</details>

Reply:
This is the key part in `Test2`: `valid_range: 0..=2,`

So, 1 *is* a valid discriminant.